plugins/pulp_rpm/plugins/importers/yum/purge.py

```diff
@@ -286,3 +287,106 @@ def remove_unit_duplicate_nevra(unit, repo):
+                continue
+
+            repo = model.Repository.objects.get(repo_id=repo_id)
+
+            # Since the repo_units queryset is ordered by the updated field (descending), the
+            # first repo content unit is the latest. All other RCUs should be disassociated
+            duplicate_units = (unit_type(id=rcu.unit_id) for rcu in rcus[1:])
+            repo_controller.disassociate_units(repo, duplicate_units)
+
+
+def _duplicate_key_id_generator(unit):
+    """duplicate NEVRA unit ID generator
+
+    :param unit: The unit whose NEVRA should be removed
+    :type unit: ContentUnit
+
+    find all duplicate NEVRAs, regardless of repository, for a given content unit type
+
+    This is a memory-efficient pre-filter, which finds all possible duplicate NEVRAs in the
+    collection of a content unit that supports the NEVRA fields. This iterator can then be
+    cross-referenced with a repository to remove duplicate repository content unit
+    links for a given NEVRA.
+    """
+    # In order to find potential duplicate nevras, a mongo mapReduce is employed to find units
+    # with duplicate nevra in a unit's entire collection based on non-checksum unit key fields.
+    # Filtering with QuerySet overflows mongo's sort stage buffer, and doing it purely with
+    # python uses too much memory. Running things through mapReduce results in an iterable that
+    # makes it relatively simple to spit out potential duplicate unit IDs without
+    # breaking either mongo or python.
+
+    # consider duplicates to be units with the same unit key when the checksum is ignored
+    # normally that means NEVRA, but with DRPMs, for example, filename is used instead of name
+    fields = list(unit.NAMED_TUPLE._fields)
+    fields.remove('checksum')
+    fields.remove('checksumtype')
+
+    if unit is models.DRPM:
+        name_field = 'filename'
+    else:
+        name_field = 'name'
+
+    # going with % interpolation here because all the {}'s make .format no fun for JS
+    # map all units in the collect into key value pairs, where the key is the unit
+    # key fields joined by '-', and the value is the unit id
+    map_f = """
+    function () {
+        var key_fields = [this.%s, this.epoch, this.version, this.release, this.arch]
+        emit(key_fields.join('-'), {ids: [this._id]});
+    }
+    """ % name_field
+
+    # reduce keys with multiple values down by collecting ids into a single return object
+    reduce_f = """
+    function (key, values) {
+      // collect mapped values into the first value to build the list of ids for this key/nevra
+      var collector = values[0]
+      // since collector is values[0] start this loop at index 1
+      // reduce isn't called if map only emits one result for key,
+      // so there is at least one value to collect
+      for (var i = 1; i < values.length; i++) {
+        collector.ids = collector.ids.concat(values[i].ids)
+      }
+      return collector
+    }
+    """
```

Comment:
Haha, someone just wrote pulp-list today to ask if we use the MongoDB JS engine. So I guess you can write back with a "we do nooooow!" or similar. ☺